### PR TITLE
[Lang] MatrixNdarray refactor part4: Lowered TensorType to CHI IR level for elementwise-indexed MatrixNdarray

### PR DESCRIPTION
**Pull request by jim19930609**
Related issue = https://github.com/taichi-dev/taichi/issues/5873, https://github.com/taichi-dev/taichi/issues/5819

This PR is working on "Part ②", "Part ③" and "Part ④" in https://github.com/taichi-dev/taichi/issues/5873.

[Overview]
When "real_matrix = True", suppose we have the following code with elementwise-indexed MatrixNdarray involved:

```
@ti.kernel
def func(a: ti.types.ndarray()):
    for i in range(5):
        for j in range(4):
            a[i][j * j] = j * j 

v = ti.Vector.ndarray(10, ti.i32, 5, layout=ti.AOS)
func(v)
```

The lowering pass will be:

NdarrayMatrix_Indexing drawio

[Major changes]
1. Change on Frontend Language
In `python/taichi/lang/impl.py`, we modified `subscript` methods to turn-off scalarization for ExternalTensorExpression. 

2. Change on IR (with real_matrix = True):
- ArgLoadStmt: Now able to return TensorType
- PtrOffsetStmt: Now able to take `ExternalPtrStmt` as base pointer
- IndexExpression: Now able to index `ExternalTensorExpression` 

3. Change on CodeGen
Adjusted codegen logic for ArgLoadStmt, PtrOffsetStmt and ExternalPtrStmt to correctly handle TensorType

### CODE REVIEW

**Review by strongoier — APPROVED** [review]
Great!! Only some nits.

File: taichi/ir/frontend_ir.h [hunk 1]
```diff
@@ -465,9 +465,6 @@ class ExternalTensorExpression : public Expression {
     }
   }
 
-  void type_check(CompileConfig *config) override {
```
[the remark this answers is not in the document]
Comment (jim19930609) [reply taichi/ir/frontend_ir.h:468]: Thanks! For this change, let me localize `config` to `ExternalTensorExpression` only. We can promote it to a member of `Expression` in the following PRs if needed~

File: taichi/transforms/compile_to_offloads.cpp [hunk 2]
```diff
@@ -256,6 +256,7 @@ void offload_to_executable(IRNode *ir,
   irpass::full_simplify(
       ir, config,
       {lower_global_access, /*autodiff_enabled*/ false, kernel->program});
+
```
[the remark this answers is not in the document]
Comment (jim19930609) [reply taichi/transforms/compile_to_offloads.cpp:259]: thx~

File: taichi/python/export_lang.cpp [hunk 3]
```diff
@@ -739,6 +739,8 @@ void export_lang(py::module &m) {
            [](Expr *expr) { return expr->is<GlobalVariableExpression>(); })
       .def("is_external_var",
            [](Expr *expr) { return expr->is<ExternalTensorExpression>(); })
+      .def("is_index_var",
```
[the remark this answers is not in the document]
Comment (jim19930609) [reply taichi/python/export_lang.cpp:742]: Sure. For the other ones, let me leave a TODO here for now.

File: python/taichi/lang/impl.py [hunk 4]
```diff
@@ -224,6 +225,16 @@ def subscript(value, *_indices, skip_reordered=False, get_ref=False):
...
         ret.any_array_access = any_array_access
         return ret
+    if isinstance(value, Expr):
+        # Index into TensorType
+        # value: IndexExpression with ret_type = TensorType
+        assert current_cfg().real_matrix is True
+        assert value.ptr.is_index_var()
+        assert is_tensor(value.ptr.get_ret_type())
+        return Expr(
+            _ti_core.subscript(value.ptr, indices_expr_group,
```
[the remark this answers is not in the document]
Comment (jim19930609) [reply python/taichi/lang/impl.py:235]: Let me leave a TODO here and merge these two functions in a following PR~

File: taichi/python/export_lang.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jim19930609) [reply taichi/python/export_lang.cpp]: https://github.com/taichi-dev/taichi/pull/5991

[the remark this answers is not in the document]
Comment (strongoier) [reply taichi/python/export_lang.cpp]: Oops. It seems that I already covered #5991 in #5989 lol

[the remark this answers is not in the document]
Comment (jim19930609) [reply taichi/python/export_lang.cpp]: great!

**Pull request merged**